File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/memento/MementoPattern.java
package jv.com.tencent.cain.design_pattern.memento;

/**
 * @author : jiangyu
 * @date : 2020/4/17
 * @desc : 备忘录设计模式
 */
public class MementoPattern {
    public static void main(String[] args) {
        Originator originator = new Originator();
        CareTaker careTaker = new CareTaker();
        originator.setState("###state 1###");
        careTaker.addMomento(originator.saveStateMemento());
        originator.setState("###state 2###");
        careTaker.addMomento(originator.saveStateMemento());
        originator.setState("###state 3###");
        careTaker.addMomento(originator.saveStateMemento());

        originator.setStateMemento(careTaker.getMemento(0));
        System.out.println("cur state: " + originator.getState());
        originator.setStateMemento(careTaker.getMemento(1));
        System.out.println("cur state: " + originator.getState());
        originator.setStateMemento(careTaker.getMemento(2));
        System.out.println("cur state: " + originator.getState());
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/memento/MementoPattern.kt
package kt.com.tencent.cain.desigin_pattern.memento

/**
 * @author : jiangyu
 * @date   : 2020/4/17
 * @desc   : 备忘录设计模式
 */
fun main() {
    val originator = Originator()
    val careTaker = CareTaker()
    originator.state = "---state 1---"
    careTaker.addMemento(originator.saveStateMemento())
    originator.state = "---state 2---"
    careTaker.addMemento(originator.saveStateMemento())
    originator.state = "---state 3---"
    careTaker.addMemento(originator.saveStateMemento())

    careTaker.getMemento(0)?.let {
        originator.setStateMemento(it)
        println("cur state: ${originator.state}")
    }
    careTaker.getMemento(1)?.let {
        originator.setStateMemento(it)
        println("cur state: ${originator.state}")
    }
    careTaker.getMemento(2)?.let {
        originator.setStateMemento(it)
        println("cur state: ${originator.state}")
    }
}